### ISSUE TITLE
Expose target service name in middleware

### ISSUE THREAD

**Issue by WDaan** [traefik/traefik#10427]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

I'm writing a logger plugin for Traefik, it will log information like request duration, path, status code, ...
The goal is to use this middleware as a unified place to capture logs from Traefik, as it's being used as an API Gateway.

One of the things I would like to log is the 'service' the request is meant to be for.

The only place where I found this information is when I logged the `next` request handler.

`os.Stdout WriteString(fmt.Sprintf"%+v\n", next))`

gives me

`&{router:whoamiadocker service:whoami-traefik-logger next:0x4000743f0}`

Currently I use this hacky way to extract the value nonetheless.

```go
// New instantiates and returns the required components used to handle an HTTP request.
func New(_ context.Context, next http.Handler, config *Config, name string) (http.Handler, error) {
	service := reflect.ValueOf(next).Elem().FieldByName("service")
....
```

Surely, there must be a better way to do this?
I couldn't find anything in the plugin guide and basically scoured exisiting plugins in the hope to find a way to tackle this.

**Comment by WDaan** [comment]
I want to run Traefik in a microservice architecture, all request will flow through Traefik.

Instead of having all microservices implement their own logger middleware, I'd like to remove that responsibility from them and move it to a Traefik plugin.

The plugin would parse usefull information from the request, obfuscate headers where needed and log the result so it can be picked up by our logging setup.

On key part of usefull information would be the service the request would get routed too.

Does that clarify it enough?

**Comment by WDaan** [comment]
Hi @mmatur 

Any further comment?

Thanks

**Comment by mmatur** [comment]
Hi @WDaan,

Why are not using the [current access log filtering](https://doc.traefik.io/traefik/observability/access-logs/#filtering) available in Traefik?

You can use [json](https://doc.traefik.io/traefik/observability/access-logs/#format) format and you will have a bunch of field available and the one you are looking for is `ServiceName`

<img width="1317" alt="image" src="https://github.com/traefik/traefik/assets/1776972/6683e07d-2a8a-4a46-a10c-0b3657c11d80">

**Comment by WDaan** [comment]
Hi @mmatur, thanks for your reply 😄 

TLDR; it's great but just not flexible enough. Traefik v3 breaks the above mentioned 'service'-hack

- Can't make fields 'top level'
   - x-amzn-trace-id => I want this to be a top level field, because downstream services include this in their log messages and during debugging you can easily filter on this field
   - user_id,org_id => Same reason, currently I extract those from the JWT (I can probably set them as headers in the forward-auth middleware, but still cannot get them top level)
- Custom log level => I set a log level based on status code or some other factor, so I can dashboard/report on that
- Custom log format => our Fluentbit setup drops messages that do not adhere to certain rules. So I'd either have to make an exception for Traefik (which would break some dashboards) or massage/reformat it on Fluentbit side to support our format
- No servicename remapping => I remap the service name from `namespace-my-service-80@kubernetes` to `my-service`, so I can see all log messages (app + requests) with the same filter e.g. `service=my-service`
- No Path/Query Param extraction => `api/organizations/392dcb9d-d855-4447-97bc-3255eab8e37a/?test=1234` gets remapped to `api/organizations/:org` as path and the params get logged as a separate JSON top level fields `params` and `query`. We use this for debugging and dashboarding/analytics use-cases

It was great to be able to use a Traefik middleware to generate this, instead of relying on all downstream dev teams to implement the standard properly in various languages. The middleware removes that pain point altogether.

Similar issues for metrics; the current metrics do not include a 'cleaned' path label (or any path label). We use this label so we can split our SLO's based on paths we expect to be fast/slow. Or on other labels that can be set from within the middleware. We currently export metrics ourselves because the default provided ones are not extensive enough and other labels can only be added from headers. On of the labels we set is the remapped service name.

Now with Traefik v3, this setup is broken because the 'service' field is no longer exists on the `http.Handler`

I tried to extract the `LogData` from the context like the `accesslog`-middleware does it but was unable get anything.
```go
func GetLogData(req *http.Request) *LogData {
	if ld, ok := req.Context().Value(DataTableKey).(*LogData); ok {
		return ld
	}
	return nil
}
```

Is there any other way of getting access to the service name from within a middleware?

Thanks for your input 😄
